File: leetCode.com/typescript/strings/validAnagram.ts
// Given two strings s and t, return true if t is an anagram of s, and false otherwise.

function isAnagram1(s: string, t: string): boolean {
    if (s.length != t.length)
        return false

    let count: { [key: string]: number } = {}
    for (const c of s)
        count[c] = (count[c] ?? 0) + 1
    for (const c of Object.keys(count)) {
        if (count[c] == Array.from(t.matchAll(new RegExp(c, 'g'))).length)
            continue  
        else
            return false
    }
    return true
};

// Or
function isAnagram2(s: string, t: string): boolean {
    if (s.length != t.length)
        return false

    let count: { [key: string]: number } = {}
    for (const c of s)
        count[c] = (count[c] ?? 0) + 1
    for (const c of t) {
        if (!count[c])
            return false
        count[c] -= 1
    }
    return true
};

File: leetCode.com/python/strings/validAnagram.py
# Given two strings s and t, return true if t is an anagram of s, and false otherwise.

def isAnagram1(s: str, t:str):
    if len(s) != len(t):
        return False
    count = {}
    for i, c in enumerate(s):
        count[c] = count.get(c, 0) + 1
    for k in count:
        if t.count(k) != count.get(k):
            return False
    return True

# Or
def isAnagram2(s: str, t:str):
    if len(s) != len(t):
        return False
    count = {}
    for c in s:
        count[c] = count.get(c, 0) + 1
    for c in t:
        if c not in count or count.get(c) == 0:
            return False
        count[c] -= 1
    return True
    